targets/csharp/source/PlayFabSDK/source/Qos/PlayFabQosApi.cs

```diff
@@ -0,0 +1,139 @@
+        }
+
+        private async Task InitializeQoSServerList()
+        {
+            if (_dataCenterMap.Count > 0)
+            {
+                // If the dataCenterMap is already initialized, return
+                return;
+            }
+
+            var request = new ListQosServersRequest();
+            PlayFabResult<ListQosServersResponse> response = await PlayFabMultiplayerAPI.ListQosServersAsync(request);
+            
+            if (response == null || response.Error != null)
+            {
+                return;
+            }
+
+            foreach (QosServer qosServer in response.Result.QosServers)
+            {
+                if (!qosServer.Region.HasValue)
+                {
+                    continue;
+                }
+
+                _dataCenterMap[qosServer.Region.Value] = qosServer.ServerUrl;
+            }
+        }
+
+        private async Task<List<QosRegionResult>> GetSortedRegionLatencies(int timeoutMs)
+        {
+            var asyncPingResults = new List<Task<QosRegionResult>>(_dataCenterMap.Count);
+            foreach (KeyValuePair<AzureRegion, string> datacenter in _dataCenterMap)
+            {
+                var regionPinger = new RegionPinger(datacenter.Value, datacenter.Key);
+                Task<QosRegionResult> pingResult = regionPinger.PingAsync(timeoutMs);
+                asyncPingResults.Add(pingResult);
+            }
+
+            await Task.WhenAll(asyncPingResults);
+            var results = new List<QosRegionResult>(asyncPingResults.Count);
+            foreach (Task<QosRegionResult> asyncPingResult in asyncPingResults)
+            {
+                results.Add(asyncPingResult.Result);
+            }
+
+            results.Sort((x,y) => x.LatencyMs.CompareTo(y.LatencyMs));
+
+            return results;
+        }
+
+        private async Task SendResultsToPlayFab(QosResult result)
+        {
+            var eventContents = new EventContents
+            {
+                Name = "qos_result",
+                EventNamespace = "playfab.servers",
+                Payload = QosResultFacade.CreateFrom(result)
+            };
+
+            var writeEventsRequest = new WriteEventsRequest
+            {
+                Events = new List<EventContents> {eventContents}
+            };
+
+            await WriteTelemetryAsync(writeEventsRequest);
+        }
+
+        private static async Task<PlayFabResult<WriteEventsResponse>> WriteTelemetryAsync(WriteEventsRequest request, object customData = null, Dictionary<string, string> extraHeaders = null)
+        {
+            if ((request?.AuthenticationContext?.EntityToken ?? PlayFabSettings.staticPlayer.EntityToken) == null) throw new PlayFabException(PlayFabExceptionCode.EntityTokenNotSet, "Must call GetEntityToken before calling this method");
+
+            var httpResult = await PlayFabHttp.DoPost("/Event/WriteTelemetryEvents", request, "X-EntityToken", PlayFabSettings.staticPlayer.EntityToken, extraHeaders);
```

Comment:
It's unusual to call obscured APIs directly with PlayFabHttp.
Is this not a public API method?
Note, that by definition, releasing open source client code that calls it, makes it public anyway.

Reply:
As we chatted. This is already in Beta. We can change it to call the real API later once it is in prod. 

---
In reply to: [291755067](https://github.com/PlayFab/SDKGenerator/pull/535#discussion_r291755067) [](ancestors = 291755067)